src/gil.rs

```diff
@@ -187,42 +176,65 @@ impl ReleasePool {
     }
 }
 
-static mut POOL: *mut ReleasePool = ::std::ptr::null_mut();
+struct ReleasePool {
+    value: UnsafeCell<Option<ReleasePoolImpl>>,
+}
+
+impl ReleasePool {
+    const fn new() -> Self {
+        Self {
+            value: UnsafeCell::new(None),
+        }
+    }
+    unsafe fn get_or_init(&self) -> &mut ReleasePoolImpl {
+        if (*self.value.get()).is_none() {
+            *self.value.get() = Some(ReleasePoolImpl::new());
+        }
+        (*self.value.get()).as_mut().unwrap()
+    }
+}
+
+static POOL: ReleasePool = ReleasePool::new();
+
+unsafe impl Sync for ReleasePool {}
 
 #[doc(hidden)]
-pub struct GILPool<'p> {
-    py: Python<'p>,
+pub struct GILPool {
     owned: usize,
     borrowed: usize,
     no_send: Unsendable,
 }
```

Comment:
Now that GILPool doesn't have `py`, I think we can change GILGuard to be 

```
struct GILGuard {
    gstate: ffi::PyGILState_State,
    pool: GILPool
}
```

and then we can simplify _a lot_ of the duplication in this file.

Reply:
Seems nice :+1: